cli/flox/src/utils/search.rs

```diff
@@ -87,108 +87,148 @@ pub(crate) fn construct_search_params(
     Ok(params)
 }
 
-/// Deduplicate and disambiguate display items.
+/// An intermediate representation of a search result used for rendering
+#[derive(Debug, PartialEq, Clone)]
+pub struct DisplayItem {
+    /// The input that the package came from
+    input: String,
+    /// The attribute path of the package, excluding subtree and system
+    rel_path: Vec<String>,
+    /// The package description
+    description: Option<String>,
+    /// Whether to join the `input` and `package` fields with a separator when rendering
+    render_with_input: bool,
+}
+
+impl Display for DisplayItem {
+    /// Render a display item in the format that should be output by
+    /// `flox search`.
+    ///
+    /// It should be possible to copy and paste this as an argument to
+    /// `flox install`.
+    ///
+    /// If we change this function, we will likely need to update what the
+    /// deduplicate field controls in pkgdb.
+    /// Technically, pkgdb shouldn't have knowledge of this format,
+    /// but it's nicer to perform deduplication in SQL.
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.render_with_input {
+            write!(f, "{}{SEARCH_INPUT_SEPARATOR}", self.input)?;
+        }
+        write!(f, "{}", self.rel_path.join("."))
+    }
+}
+
+/// Contains [DisplayItem]s that have been disambiguated.
 ///
-/// This gets complicated because we have to satisfy a few constraints:
-/// - The order of results from `pkgdb` is important (best matches come first),
-///   so that order must be preserved.
-/// - Versions shouldn't appear in the output, so multiple package versions from a single
-///   input should be deduplicated.
-/// - Packages that appear in more than one input need to be disambiguated by prepending
-///   the name of the input and a separator.
-fn dedup_and_disambiguate_display_items(mut display_items: Vec<DisplayItem>) -> Vec<DisplayItem> {
-    let mut package_to_inputs: HashMap<String, HashSet<String>> = HashMap::new();
-    for d in display_items.iter() {
-        // Build a collection of packages and which inputs they are seen in so we can tell
-        // which packages need to be disambiguated when rendering search results.
-        package_to_inputs
-            .entry(d.package.clone())
-            .and_modify(|inputs| {
-                inputs.insert(d.input.clone());
+/// This should be used for printing search results when the format output by
+/// [DisplaySearchResults] is not desired.
+pub struct DisplayItems(pub Vec<DisplayItem>);
```

Comment:
suggestion: either derive `derive_more::Deref` or add `iter()` as a inherent impl.
exposing inner fields of newtypes is considered leaky.


```suggestion
pub struct DisplayItems(Vec<DisplayItem>);
```

```rust
fn iter(&self) -> impl Iterator<Item=DisplayItem> {
    self.0.iter()
}
```

Reply:
Fixing